src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesAuthFactory.java

```diff
@@ -0,0 +1,106 @@
+package org.csanchez.jenkins.plugins.kubernetes;
+
+import com.cloudbees.plugins.credentials.CredentialsMatchers;
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.cloudbees.plugins.credentials.common.StandardCertificateCredentials;
+import com.cloudbees.plugins.credentials.common.StandardCredentials;
+import com.cloudbees.plugins.credentials.common.UsernamePasswordCredentials;
+import hudson.security.ACL;
+import jenkins.authentication.tokens.api.AuthenticationTokens;
+import jenkins.model.Jenkins;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.io.IOUtils;
+import org.jenkinsci.plugins.docker.commons.credentials.DockerServerCredentials;
+import org.jenkinsci.plugins.kubernetes.credentials.TokenProducer;
+import org.jenkinsci.plugins.plaincredentials.FileCredentials;
+import org.jenkinsci.plugins.plaincredentials.StringCredentials;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyStore;
+import java.security.cert.X509Certificate;
+import java.util.Collections;
+
+import org.jenkinsci.plugins.kubernetes.auth.*;
+
+public abstract class KubernetesAuthFactory {
+
+    private static StandardCredentials getCredentials(String credentialsId) {
+        return CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(
+                        StandardCredentials.class,
+                        Jenkins.get(),
+                        ACL.SYSTEM,
+                        Collections.emptyList()
+                ),
+                CredentialsMatchers.withId(credentialsId)
+        );
+    }
+
+    private static StandardCredentials getAuthenticationTokenCredentials(String credentialsId) {
+        return CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(
+                        StandardCredentials.class,
+                        Jenkins.get(),
+                        ACL.SYSTEM,
+                        Collections.emptyList()
+                ),
+                CredentialsMatchers.allOf(
+                        AuthenticationTokens.matcher(KubernetesAuth.class),
+                        CredentialsMatchers.withId(credentialsId)
+                )
+        );
+    }
+
+    public static KubernetesAuth fromCredentialsId(String credentialsId, String serviceAddress, String caCertData, Boolean skipTlsVerify) throws Exception {
+        StandardCredentials c = getAuthenticationTokenCredentials(credentialsId);
+        if (c != null) {
+            return AuthenticationTokens.convert(KubernetesAuth.class, c);
+        }
+        c = getCredentials(credentialsId);
+        if (c instanceof StringCredentials) {
```

Comment:
All this `if...else` sequence could be replaced by implementations of `AuthenticationTokenSource`

Reply:
Surely it could (and should) be, but as we decided in https://github.com/jenkinsci/kubernetes-plugin/pull/580, this should be PR with initial changes to support `AuthenticationTokenSource`. I thought it would be better to split that work at least here to prevent too big PRs

Reply:
Actually by minimal I thought you wouldn't convert existing implementations over to `KubernetesAuth`.

Reply:
It is my first contribution to Jenkins code and I may be not so good with Java and its patterns, but I thought it would be better way to start moving to new API instead of keeping diversed methods for obtaining auth data.

Reply:
This is completely fine. Thank you for looking at it!